File: p034.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        l = 0
        r = len(nums)
        if r == 0: return [-1, -1]
        while r-l > 1:
            m = (r+l)//2
            if nums[m] < target: l = m
            else: r = m
        result = l
        if nums[result] < target:
            result += 1
            if result >= len(nums) or nums[result] != target: return [-1, -1]
        elif nums[result] > target: return [-1, -1]
        l = result
        r = len(nums)
        while r-l > 1:
            m = (r+l)//2
            if nums[m] > target: r = m
            else: l = m
        return [result, l]
            

File: p034.c
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* searchRange(int* nums, int numsSize, int target, int* returnSize) {
    int *result = (int*)malloc(sizeof(int)*2);
    *returnSize = 2;
    unsigned l = 0;
    unsigned r = numsSize;
    if (!r)
    {
        result[0] = result[1] = -1;
        return result;
    }
    unsigned rr = r;
    unsigned m;
    while (r-l > 1)
    {
        m = (r+l)>>1;
        if (nums[m] < target)
        {
            l = m;
        }
        else
        {
            r = m;
            if (nums[m] > target) rr = m;
        }
    }
    result[0] = l;
    if (nums[result[0]] < target)
    {
        result[0]++;
        if (result[0] >= numsSize || nums[result[0]] != target)
        {
            result[0] = result[1] = -1;
            return result;
        }
    }
    else if (nums[result[0]] > target)
    {
        result[0] = result[1] = -1;
        return result;
    }
    l = result[0];
    r = rr;
    while (r-l > 1)
    {
        m = (r+l)>>1;
        if (nums[m] > target)
        {
            r = m;
        }
        else
        {
            l = m;
        }
    }
    result[1] = l;
    return result;
}